Bluetooth: audio: Add security requirements check

MariuszSkamra opened the pull request:
The LE Audio related profiles define specific security requirements that
client have to met to gain access to related characteristics. This
introduces internal wrapper API for GATT characteristics and CCC
descriptors to ensure the security.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: keys: Extend keys flags with OOB flag

Adds flag indicating the pairing that was completed using OOB method.
The flag is needed for checking 128bit entropy pairing required by LE
Audio services.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2651,7 +2651,9 @@ uint8_t bt_gatt_check_perm(struct bt_conn *conn, const struct bt_gatt_attr *attr
-		if ((conn->le.keys->flags & BT_KEYS_SC) == 0) {
+		if ((conn->le.keys->flags & BT_KEYS_SC) == 0 &&
+		    ((conn->le.keys->flags & BT_KEYS_OOB) == 0 ||
+		     bt_conn_enc_key_size(conn) < 16)) {
```
`BT_SMP_MAX_ENC_KEY_SIZE` instead of `16`

Thalley reviewed (dismissed):
Minor request, otherwise LGTM.

jori-nordic commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -72,6 +72,32 @@ enum {
 	 */
 	BT_GATT_PERM_PREPARE_WRITE = BIT(6),
+
+	/** @brief Attribute write permission with 128-bit entropy key connection.
```
read

Thalley replied:
This change is from https://github.com/zephyrproject-rtos/zephyr/pull/46923

@MariuszSkamra I guess this should probably have been marked as a draft. I'll add the DNM until https://github.com/zephyrproject-rtos/zephyr/pull/46923 is merged.

jori-nordic commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -72,6 +72,32 @@ enum {
+
+	/** @brief Attribute write permission with 128-bit entropy key connection.
+	 *
+	 *  If set, write access requires connection using an at least 128 bits of
```
read

MariuszSkamra committed:
Bluetooth: conn: Add const modifier to conn

The bt_conn_get_security and bt_conn_enc_key_size do not
modify the conn thus the parameter can be guarded with const modifier.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: conn: Move bt_security_t definition

This moves bt_security_t definition to be used later in bt_conn_info.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: conn: Extend bt_conn_info with security information

This extends the bt_conn_info with security information. This is needed
by LE Audio profiles to check if security requirements are met.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
doc: release-3.2: Document bt_conn_get_security and bt_conn_enc_key_size

Note on const modifier added to `conn` parameter.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

fredrikdanebjer commented on subsys/bluetooth/host/smp.c:
```diff
@@ -5750,16 +5750,19 @@ void bt_smp_update_keys(struct bt_conn *conn)
+	case LEGACY_OOB:
+		conn->le.keys->flags |= BT_KEYS_OOB;
+		/* fallthrough */
 	case PASSKEY_DISPLAY:
```
Seeing as how the discussion in [#46923](https://github.com/zephyrproject-rtos/zephyr/pull/46923) is coming along, maybe we should differentiate between these two. We can then check just `BT_KEYS_LE_SC_OOB` for the `LESC` permission.

Its possible that a 128B permission is still needed to be (re-)added, to cover the Legacy OOB with 128B entropy key for LE Audio though.

MariuszSkamra replied:
Could you please check the current revision?

fredrikdanebjer replied:
The above mentioned issue is solved by the le-audio security check that you implemented. I think this is the best solution too!

Thalley commented:
Since the 128B addition was reject in https://github.com/zephyrproject-rtos/zephyr/pull/46923 by this comment github.com/zephyrproject-rtos/zephyr/pull/46923#issuecomment-1180111659 I think you need to provide the reason why you are trying to add it here again, and what changed.

alwa-nordic commented:
I created a separate PR (https://github.com/zephyrproject-rtos/zephyr/pull/48165) for the bugfix in `bt_gatt_check_perm`. When it's in and you rebase this branch, the commit on this branch should be recognized as duplicate and automatically removed.

alwa-nordic commented:
I am OK with adding the `BT_KEYS_OOB` flag.

I am not happy with the `BT_GATT_PERM_READ_128B`/`BT_GATT_PERM_WRITE_128B` flags because it's not at all clear what it means. I don't think we have selected an objective definition of what it means for a link to have '128 bit entropy key encryption'. LESC JustWorks connection under a MITM attack is completely readable by the attacker. Yet it fulfills the requirement. A link secured with Legacy pairing (even using JustWorks) that was not sniffed during pairing and has a 128b LTK key is as safe as LESC. Once the LTK is established, the encryption mechanism is the same. Yet it fails this requirement.

The closest meaningful promise I can think of is 'Key exchange was safe from passive sniffing, (but not necessarily from MITM)'. That assumes the OOB implementation cannot be sniffed. And that may be a real attack vector if NFC is used.

Since this is an Audio spec definition, I think the implementation should go in that layer. The read and write handlers of the services can return `BT_ATT_ERR_AUTHENTICATION`. I see `BT_GATT_PERM` as a convenience, not a one-size-fits-all.

MariuszSkamra committed:
Bluetooth: audio: Add security requirements check

The LE Audio related profiles define specific security requirements that
client have to met to gain access to related characteristics. This
introduces internal wrapper API for GATT characteristics and CCC
descriptors to ensure the security.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
> I am OK with adding the `BT_KEYS_OOB` flag.
> 
> I am not happy with the `BT_GATT_PERM_READ_128B`/`BT_GATT_PERM_WRITE_128B` flags because it's not at all clear what it means. I don't think we have selected an objective definition of what it means for a link to have '128 bit entropy key encryption'. LESC JustWorks connection under a MITM attack is completely readable by the attacker. Yet it fulfills the requirement. A link secured with Legacy pairing (even using JustWorks) that was not sniffed during pairing and has a 128b LTK key is as safe as LESC. Once the LTK is established, the encryption mechanism is the same. Yet it fails this requirement.
> 
> The closest meaningful promise I can think of is 'Key exchange was safe from passive sniffing, (but not necessarily from MITM)'. That assumes the OOB implementation cannot be sniffed. And that may be a real attack vector if NFC is used.
> 
> Since this is an Audio spec definition, I think the implementation should go in that layer. The read and write handlers of the services can return `BT_ATT_ERR_AUTHENTICATION`. I see `BT_GATT_PERM` as a convenience, not a one-size-fits-all.

Done. Could you please take a look at current revision?

MariuszSkamra committed:
Bluetooth: audio: Keep ASE pointer in bt_audio_ep

This makes the server endpoint to store the ASE pointer instead of
handle. This way, there will be no need to "mock" attribute when
notification is sent.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley reviewed (changes requested):
The change to ASCS and ASE pointer seems unrelated to this PR, and should be a in separate PR IMO.

MariuszSkamra commented:
> The change to ASCS and ASE pointer seems unrelated to this PR, and should be a in separate PR IMO.

It is related, as it touches the user-data I'm changing in the following PR

MariuszSkamra commented on subsys/bluetooth/audio/aics.c:
```diff
@@ -65,36 +66,33 @@ static ssize_t read_description(struct bt_conn *conn,
-		    BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_LESC)               \
+	BT_AUDIO_CHRC(BT_UUID_AICS_STATE,                                      \
+		      BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,                 \
+		      BT_GATT_PERM_READ_ENCRYPT,                               \
```
neither `BT_GATT_PERM_READ_LESC` nor `BT_GATT_PERM_WRITE_LESC` are needed, because the access to the characteristic is validated on service level with this patch.

Thalley replied:
Right, but if we set them "correctly" they could be validated at an earlier time at the GATT level instead. It doesn't really make much sense to postpone the LESC check to later (in audio.c) when it can just as easily be done earlier (in gatt.c), which would also make the characteristic declaration better matter the actual security check.

MariuszSkamra replied:
But wouldn't `BT_GATT_PERM_WRITE_LESC` exclude OOB method? If I understand correctly, setting `BT_GATT_PERM_WRITE_LESC` would reject access to attribute if not paired with LE Secure Connections.

fredrikdanebjer replied:
I'm in agreement with you @MariuszSkamra, the `BT_GATT_PERM_WRITE_LESC` would indeed exclude the OOB method.
Furthermore, the OOB method is a very special case (both key size + OOB), which we as a combination are unlikely to get into `gatt.c`.

I think the current approach is the best

Thalley replied:
Ah, right, I see. Good point. I had no considered that adding `BT_GATT_PERM_WRITE_LESC` would prevent OOB pairing in that regard. 

I am also in agreement of this.

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -2273,55 +2266,46 @@ static ssize_t ascs_cp_write(struct bt_conn *conn,
-		    BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT),
+	BT_AUDIO_CHRC(BT_UUID_ASCS_ASE_SNK,
+		      BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
+		      BT_GATT_PERM_READ_ENCRYPT,
```
Nope, see my comment

MariuszSkamra commented on subsys/bluetooth/audio/audio.c:
```diff
@@ -0,0 +1,99 @@
+/* Common functions for LE Audio services */
```
Sorry, but as discussed, no one wants the `128B` flag anymore to be in the GATT API. Thus the service has to validate the additional security requirements on service level in GATT Read/Write callback context. The good thing in this approach s that the security is checked in one place. IMO it's advantage, rather than disadvantage.

Another approach I was thinking of was to have macro/function testing the security. This would work for `BT_GATT_CHARACTERISTIC` but for `BT_GATT_CCC` we would still need to use `BT_GATT_CCC_MANAGED` instead with `cfg_write` validating the security. This, I decided that it's cleaner to `override` the GATT API.

Thalley replied:
The CCC check is a good reason for this approach :)

jori-nordic commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2263,6 +2263,13 @@ int bt_conn_get_info(const struct bt_conn *conn, struct bt_conn_info *info)
 	info->role = conn->role;
 	info->id = conn->id;
 	info->state = conn_internal_to_public_state(conn->state);
+	info->security.flags = 0;
```
I'd prefer if the flags = 0 was moved right before setting the two bits @L2294

MariuszSkamra commented:
There's compliance check failing on this PR:
```
-:24: WARNING:NEW_TYPEDEFS: do not add new typedefs
#[24](https://github.com/zephyrproject-rtos/zephyr/runs/7538294382?check_suite_focus=true#step:9:25): FILE: include/zephyr/bluetooth/conn.h:327:
+typedef enum __packed {
```

But this does not add any new typedefs, it just moves `bt_security_t`

Thalley commented:
> There's compliance check failing on this PR:
> 
> ```
> -:24: WARNING:NEW_TYPEDEFS: do not add new typedefs
> #[24](https://github.com/zephyrproject-rtos/zephyr/runs/7538294382?check_suite_focus=true#step:9:25): FILE: include/zephyr/bluetooth/conn.h:327:
> +typedef enum __packed {
> ```
> 
> But in fact I did not add any new typedefs, just moved `bt_security_t` only

I think this needs to be overwritten by someone like @carlescufi or @jhedberg

MariuszSkamra commented:
Rebased to resolve conflicts

Thalley commented:
@carlescufi @jhedberg Compliance check fails due to moving a typedef'ed existing struct. Please disregard and overwrite that CI failure an merge when possible :)
